### DISCUSSION TITLE
How to tune cutlass matmul kernels to approach cublasLt one?

### DISCUSSION THREAD

**Discussion by MARD1NO** [NVIDIA/cutlass#700] [Q&A] [answered]
Recently I want to compare cublasLt kernel and cutlass kernel. 

Actually cublasLt calls the cutlass matmul kernel, but I cannot get the same result, maybe I need to change some configuration in cutlass kernel. 

I test the matmul+bias+gelu pattern, and I found there is an example: 12_gemm_bias_relu, I just simply change to use linearCombinationGELU epilogue. 

The case is: 
```
const int length_m = 4096;
const int length_n = 5120;
const int length_k = 4096;
```

And cublasLt's kernel name is: `void cutlass::Kernel<cutlass_80_tensorop_f16_s16816gemm_gelu_f16_128x128_32x5_nn_align8>`

I found the kernel naming rules in `generators.py` and change the configure as below: 
```cpp
using ShapeMMAThreadBlock = cutlass::gemm::GemmShape<128, 128, 32>; 
using ShapeMMAWarp = cutlass::gemm::GemmShape<64, 64, 32>; 
using ShapeMMAOp = cutlass::gemm::GemmShape<16, 8, 16>; 

using SwizzleThreadBlock = cutlass::gemm::threadblock::GemmIdentityThreadblockSwizzle<8>;
constexpr int NumStages = 5; 
```

And I also found cublasLt use tanh to approximate GELU function: 
<img width="502" alt="image" src="https://user-images.githubusercontent.com/42901638/202170417-420b7118-4335-4fff-a0d8-00176129a2ec.png">

So I change to use GELU_taylor epilogue: 
```cpp
using LinearCombinationFastGELU = LinearCombinationGeneric<GELU_taylor, ElementOutput_, Count, ElementAccumulator_,
                                                       ElementCompute_, Scale, Round, true>;
```

And it still have some gap: 
cublasLt 855us vs cutlass 900us, and I also found the grid configuration is different. cublasLt is (320, 4, 2), cutlass is (320, 4, 1). And I wonder is there exists some configurations are not aligned with cublasLt？

**Comment by hwu36** [comment, marked as answer]
You are very good!

cublas has 2 in its grid.z which means it uses splitK.  To do apple to apple comparison, you need to do the same.

Note, different version of cutlass and compiler will cause some performance difference.

**Reply by MARD1NO** [reply]
Thanks! I will try it!

**Reply by MARD1NO** [reply]
Finally, I found in cuda11.6, cutlass's performance is same as cublasLt one. In previous experiment, I use cuda11.7......

And I also compare with splitK version. origin 760us vs splitK2(the same as cublasLt) 800us

**Reply by hwu36** [reply]
Yeah.  That is what you would see if you use 11.7.  Our gtc22 fall talk mentioned the choice of the compiler.

**Reply by alohali** [reply]
@hwu36 hi, do you have notes about this? " Our gtc22 fall talk mentioned the choice of the compiler." we have no idea about compiler version choice but only use tot

**Reply by hwu36** [reply]
tot is good.  next will be better.  for gemm, 11.3+ is needed, just avoid 11.7

**Reply by alohali** [reply]
cool! thanks!
